### PR TITLE
Update LibertyRuntimeTransformer and tests

### PR DESCRIPTION
**Pull request by KyleAure**
This PR is to minimally test that we can support instrumentation up to Java 17 and leave the rest of the work of flushing out long term solutions for instrumentation to future Java work.

I think the updates to the unit tests I've introduced should mean this is a safe enough change to merge, though there are likely some corner cases in Java 11, 17, and 21 that are currently being missed.

- proposal to update the LibertyRuntimeTransformer to support Java 11 and Java 17
- get work started on Java 21

### CODE REVIEW

File: dev/com.ibm.ws.ras.instrument/src/com/ibm/ws/ras/instrument/internal/main/LibertyRuntimeTransformer.java [hunk 1]
```diff
@@ -202,7 +202,7 @@ private static boolean isTransformPossible(byte[] bytes) {
         if (isJDK8WithHotReplaceBug)
             return classFileVersion <= Opcodes.V1_7;
         else
-            return classFileVersion <= Opcodes.V11;
+            return classFileVersion <= Opcodes.V17;
```
Comment (gjwatts) [inline dev/com.ibm.ws.ras.instrument/src/com/ibm/ws/ras/instrument/internal/main/LibertyRuntimeTransformer.java]: I am hoping to centralize as many of these ASM changes as possible.  Would `com.ibm.ws.org.objectweb.asm/src/io/openliberty/asm/ASMHelper.java` be available when the transformer loads?  If so, would it work to add something like a HIGHEST_JAVA_LEVEL constant to `ASMHelper.java` (or somewhere else) and use that here?

File: dev/com.ibm.ws.ras.instrument_test/test/com/ibm/ws/ras/RasTransformTest.java [hunk 2]
```diff
@@ -34,25 +39,88 @@
...
+        majorCodeMap.put("12", 56);
+        majorCodeMap.put("13", 57);
+        majorCodeMap.put("14", 58);
+        majorCodeMap.put("15", 59);
+        majorCodeMap.put("16", 60);
+        majorCodeMap.put("17", 61);
+        majorCodeMap.put("18", 62);
+        majorCodeMap.put("19", 63);
+        majorCodeMap.put("20", 64);
+        majorCodeMap.put("21", 65);
```
Comment (gjwatts) [inline dev/com.ibm.ws.ras.instrument_test/test/com/ibm/ws/ras/RasTransformTest.java:74]: Add `majorCodeMap.put("22", 66);`

File: dev/com.ibm.ws.ras.instrument_test/test/com/ibm/ws/ras/RasTransformTest.java [hunk 3]
```diff
@@ -112,92 +180,119 @@ public String unrollStack(StackTraceElement[] trace) {
 	
 	@Test
 	public void testJava6BasicInjection() {
+		assumeTrue(Integer.parseInt(System.getProperty("java.specification.version")) >= 6);
```
Comment (gjwatts) [inline dev/com.ibm.ws.ras.instrument_test/test/com/ibm/ws/ras/RasTransformTest.java:185]: Do we still need to test Java 6?  Or can we remove this test?

File: dev/com.ibm.ws.ras.instrument_test/test/com/ibm/ws/ras/RasTransformTest.java [hunk 4]
```diff
@@ -112,92 +180,119 @@ public String unrollStack(StackTraceElement[] trace) {
...
 		} catch (Exception e) {
-			System.out.println("Exception: " + e.getStackTrace().toString());
+			fail(e.getMessage());
 		}
 		
 	}
 	
 	@Test
 	public void testJava7BasicInjection() {
+		assumeTrue(Integer.parseInt(System.getProperty("java.specification.version")) >= 7);
```
Comment (gjwatts) [inline dev/com.ibm.ws.ras.instrument_test/test/com/ibm/ws/ras/RasTransformTest.java:203]: Do we still need to test Java 7?  Or can we remove this test?

File: dev/com.ibm.ws.ras.instrument_test/test/com/ibm/ws/ras/RasTransformTest.java [hunk 5]
```diff
@@ -112,92 +180,119 @@ public String unrollStack(StackTraceElement[] trace) {
...
 		} catch (Exception e) {
-			System.out.println("Exception: " + e.getStackTrace().toString());
+			fail(e.getMessage());
 		}
 		
 	}
 	
 	@Test
 	public void testJava7StaticInit() {
+		assumeTrue(Integer.parseInt(System.getProperty("java.specification.version")) >= 7);
```
Comment (gjwatts) [inline dev/com.ibm.ws.ras.instrument_test/test/com/ibm/ws/ras/RasTransformTest.java:221]: Do we still need to test Java 7?  Or can we remove this test?

File: dev/com.ibm.ws.ras.instrument_test/test/com/ibm/ws/ras/RasTransformTest.java [hunk 6]
```diff
@@ -112,92 +180,119 @@ public String unrollStack(StackTraceElement[] trace) {
...
 			transformedClass.newInstance();		
 		} catch (Exception e) {
-			System.out.println("Exception: " + e.getStackTrace().toString());
+			fail(e.getMessage());
 		}
 		
 	}
 	
-	//@Test
+//	@Test TODO this test is failing
```
Comment (gjwatts) [inline dev/com.ibm.ws.ras.instrument_test/test/com/ibm/ws/ras/RasTransformTest.java:292]: Review this test to find out why it is failing and re-add test.

File: dev/com.ibm.ws.ras.instrument_test/test/com/ibm/example/bytecode/HelloWorldJava21.java [hunk 7]
```diff
@@ -0,0 +1,21 @@
+package com.ibm.example.bytecode;
```
Comment (gjwatts) [inline dev/com.ibm.ws.ras.instrument_test/test/com/ibm/example/bytecode/HelloWorldJava21.java:13]: Add pattern Matching for switch test -> https://openjdk.org/jeps/441?

File: dev/com.ibm.ws.ras.instrument_test/test/com/ibm/example/bytecode/HelloWorldJava17.java [hunk 8]
```diff
@@ -0,0 +1,21 @@
+package com.ibm.example.bytecode;
```
Comment (gjwatts) [inline dev/com.ibm.ws.ras.instrument_test/test/com/ibm/example/bytecode/HelloWorldJava17.java:13]: Add sealed classes test -> https://openjdk.org/jeps/409?

File: dev/com.ibm.ws.ras.instrument_test/test/com/ibm/example/bytecode/HelloWorldJava11.java [hunk 9]
```diff
@@ -0,0 +1,21 @@
+package com.ibm.example.bytecode;
```
Comment (gjwatts) [inline dev/com.ibm.ws.ras.instrument_test/test/com/ibm/example/bytecode/HelloWorldJava11.java:13]: Add Local variable syntax for lambda parameters -> https://openjdk.org/jeps/323?

File: dev/com.ibm.ws.ras.instrument_test/test/com/ibm/ws/ras/RasTransformTest.java [hunk 10]
```diff
@@ -255,77 +405,74 @@ private void writeStringToFile(String toWrite, String path) {
...
-	}
+//	private void writeBytesToFile(byte[] toWrite, String path) {
+//		try {
+//			FileOutputStream fos = new FileOutputStream(path);
+//			fos.write(toWrite);
+//			fos.close();
+//		} catch (IOException e) {
+//			e.printStackTrace();
+//		}
+//	}
```
Comment (gjwatts) [inline dev/com.ibm.ws.ras.instrument_test/test/com/ibm/ws/ras/RasTransformTest.java]: Delete lines?

File: dev/com.ibm.ws.ras.instrument_test/test/com/ibm/ws/ras/RasTransformTest.java [hunk 11]
```diff
@@ -255,77 +405,74 @@ private void writeStringToFile(String toWrite, String path) {
...
+//			}
+//			transformedClass = post;
+//		} catch (Exception e) {
+//			System.out.println("Exception: " + e.getStackTrace().toString());
+//			Assert.assertTrue(false);
+//		} 
+//		
+//		return transformedClass;
+//		
+//	}
```
Comment (gjwatts) [inline dev/com.ibm.ws.ras.instrument_test/test/com/ibm/ws/ras/RasTransformTest.java]: Delete lines?

File: dev/com.ibm.ws.ras.instrument_test/test/com/ibm/ws/ras/RasTransformTest.java (hunk 3 above)
Comment (KyleAure) [reply dev/com.ibm.ws.ras.instrument_test/test/com/ibm/ws/ras/RasTransformTest.java:185]: I would say we can keep these tests even if they are not supported java levels anymore.

File: dev/com.ibm.ws.ras.instrument/src/com/ibm/ws/ras/instrument/internal/main/LibertyRuntimeTransformer.java (hunk 1 above)
Comment (KyleAure) [reply dev/com.ibm.ws.ras.instrument/src/com/ibm/ws/ras/instrument/internal/main/LibertyRuntimeTransformer.java]: That I do not know.  It would have to be tested.

**Comment by KyleAure** [comment]
#build - run a personal build
#libby - recheck headers

File: dev/com.ibm.ws.ras.instrument_test/test/com/ibm/example/bytecode/HelloWorldJava11.java (hunk 9 above)
Comment (jhanders34) [inline dev/com.ibm.ws.ras.instrument_test/test/com/ibm/example/bytecode/HelloWorldJava11.java:13]: How can these be in the `test` directory and not be built?  All classes in the test directory are going to be built as unit tests.  Also all of these java files need to have the copyright added to them.

File: dev/com.ibm.ws.ras.instrument_test/test/com/ibm/ws/ras/RasTransformTest.java (hunk 3 above)
Comment (jhanders34) [reply dev/com.ibm.ws.ras.instrument_test/test/com/ibm/ws/ras/RasTransformTest.java:185]: Just because we are doing newer Java for running, we still can have classes that have Java 6 byte code when running in the new Java levels.

File: dev/com.ibm.ws.ras.instrument/src/com/ibm/ws/ras/instrument/internal/main/LibertyRuntimeTransformer.java (hunk 1 above)
Comment (jhanders34) [reply dev/com.ibm.ws.ras.instrument/src/com/ibm/ws/ras/instrument/internal/main/LibertyRuntimeTransformer.java]: Since we support ASM 9.6 which support Java 22 and earlier, I think this should be changed to `V22` or if not that `V21`.

File: dev/com.ibm.ws.ras.instrument_test/test/com/ibm/ws/ras/RasTransformTest.java [hunk 12]
```diff
@@ -34,25 +39,89 @@
...
+        majorCodeMap.put("13", 57);
+        majorCodeMap.put("14", 58);
+        majorCodeMap.put("15", 59);
+        majorCodeMap.put("16", 60);
+        majorCodeMap.put("17", 61);
+        majorCodeMap.put("18", 62);
+        majorCodeMap.put("19", 63);
+        majorCodeMap.put("20", 64);
+        majorCodeMap.put("21", 65);
+        majorCodeMap.put("22", 65);
```
Comment (gjwatts) [inline dev/com.ibm.ws.ras.instrument_test/test/com/ibm/ws/ras/RasTransformTest.java]: Java 22 should return a value of 66

File: dev/com.ibm.ws.ras.instrument_test/test/com/ibm/ws/ras/RasTransformTest.java [hunk 13]
```diff
@@ -206,83 +308,127 @@ public void testJava8StaticInitLambdasWithInnerClass() {
...
 		}
+		
 	}
 	
-	private void writeBytesToFile(byte[] toWrite, String path) {
+	//FIXME - This test is currently failing because isTransformPossible returns false
+	//TODO expand this test to actually test Java 21 features https://openjdk.org/jeps/441
+	//@Test 
+	public void testJava21Basic() {
+		assumeTrue(Integer.parseInt(System.getProperty("java.specification.version")) >= 21);
```
Comment (jhanders34) [inline dev/com.ibm.ws.ras.instrument_test/test/com/ibm/ws/ras/RasTransformTest.java:395]: This test will also fail because we build and run unit tests with Java 17 today and Java 17 cannot run Java 21 byte code as you noted in this line.  If we wanted to keep the test and actually run it, we would have to remove the assertTrue and just return if not >= 21
```suggestion
		if (Integer.parseInt(System.getProperty("java.specification.version")) < 21) {
		    return;
		}
```

File: dev/com.ibm.ws.ras.instrument_test/test/com/ibm/ws/ras/RasTransformTest.java [hunk 14]
```diff
@@ -34,25 +39,89 @@
 
 import com.ibm.ws.ras.instrument.internal.main.LibertyRuntimeTransformer;
 
+@SuppressWarnings("deprecation")
```
Comment (jhanders34) [inline dev/com.ibm.ws.ras.instrument_test/test/com/ibm/ws/ras/RasTransformTest.java:43]: This test doesn't actually check to make sure that classes are instrumented with trace.  Should there be something added to make sure for a method there is an entry and exit and that the annotations are there?

File: dev/com.ibm.ws.ras.instrument_test/test/com/ibm/example/bytecode/HelloWorldJava11.java (hunk 9 above)
Comment (KyleAure) [reply dev/com.ibm.ws.ras.instrument_test/test/com/ibm/example/bytecode/HelloWorldJava11.java:13]: I updated the README to me more accurate.  They are compiled, but the tests doesn't use the automatically compiled classes, it will use the cached class files.

File: dev/com.ibm.ws.ras.instrument_test/test/com/ibm/ws/ras/RasTransformTest.java (hunk 13 above)
Comment (KyleAure) [reply dev/com.ibm.ws.ras.instrument_test/test/com/ibm/ws/ras/RasTransformTest.java:395]: assumeTrue does the same thing.  It will marked the test as skipped.

**Review by tbitonti — COMMENTED** [review]
I have nothing to add in addition to Jared.

File: dev/com.ibm.ws.ras.instrument_test/test/com/ibm/ws/ras/RasTransformTest.java (hunk 14 above)
Comment (KyleAure) [reply dev/com.ibm.ws.ras.instrument_test/test/com/ibm/ws/ras/RasTransformTest.java:43]: Correct, I'll leave that for someone who actually knows how this class works :)

**Comment by KyleAure** [comment]
#libby - recheck headers
#build - new build after updates

File: dev/com.ibm.ws.ras.instrument_test/test/com/ibm/ws/ras/RasTransformTest.java (hunk 13 above)
Comment (jhanders34) [reply dev/com.ibm.ws.ras.instrument_test/test/com/ibm/ws/ras/RasTransformTest.java:395]: `assumeTrue` vs `assertTrue`  Wow those are close to each other.

**Review by jhanders34 — APPROVED** [review]

File: dev/com.ibm.ws.ras.instrument/src/com/ibm/ws/ras/instrument/internal/main/LibertyRuntimeTransformer.java (hunk 1 above)
Comment (gjwatts) [reply dev/com.ibm.ws.ras.instrument/src/com/ibm/ws/ras/instrument/internal/main/LibertyRuntimeTransformer.java]: Kyle, if you like, you can leave it at a fixed value (V22 or V21) and I can work on adding the call to a new method in ASMHelper in a different PR.

**Review by gjwatts — APPROVED** [review]

**Comment by KyleAure** [comment]
#libby - recheck headers
